Libs/MRML/Core/vtkMRMLSliceNode.cxx

```diff
@@ -371,73 +408,194 @@ void vtkMRMLSliceNode::SetOrientationToReformat()
+    {
+    std::string orietation = it->first;
+    if (!orietation.compare("Reformat"))
+      {
+      continue;
+      }
+    vtkMatrix4x4* storedSliceToRAS = this->GetSliceOrientationPreset(orietation);
+    if (storedSliceToRAS != NULL)
+      {
+      if (vtkMRMLSliceNode::Matrix4x4AreEqual(storedSliceToRAS, sliceToRAS))
+        {
+        return orietation;
+        }
+      }
+    }
+
+  vtkErrorMacro("GetSliceOrientationPresetName: orientation string not found.");
+  return "";
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLSliceNode::GetSliceOrientationPresetNames(vtkStringArray *namedOrientations)
+{
+  namedOrientations->SetNumberOfValues(this->OrientationMatrices.size() + 1);
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it;
+  int id = 0;
+  for (it = this->OrientationMatrices.begin(); it != this->OrientationMatrices.end(); ++it)
+    {
+    namedOrientations->SetValue(id, it->first);
+    id++;
+    }
+  namedOrientations->SetValue(id, "Reformat");
+}
+
+//----------------------------------------------------------------------------
+bool vtkMRMLSliceNode::AddSliceOrientationPreset(const std::string &name, vtkMatrix4x4 *sliceToRAS)
+{
+  if (!name.compare("Reformat"))
+    {
+    return false;
+    }
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it = this->OrientationMatrices.find(name);
+  if (it != this->OrientationMatrices.end())
+    {
+    vtkErrorMacro("AddSliceOrientationPreset: the orietation preset " << name << " is already stored.");
+    return false;
+    }
+
+  this->OrientationMatrices[name] = sliceToRAS;
+  return true;
+}
+
+//----------------------------------------------------------------------------
+bool vtkMRMLSliceNode::RemoveSliceOrientationPreset(const std::string &name)
+{
+  if (!name.compare("Reformat"))
+    {
+    vtkErrorMacro("RemoveSliceOrientationPreset: Reformat has no orietation preset.");
+    return false;
+    }
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it = this->OrientationMatrices.find(name);
+  if (it != this->OrientationMatrices.end())
+    {
+    this->OrientationMatrices.erase(it);
+    return true;
+    }
+
+  vtkErrorMacro("RemoveSliceOrientationPreset: the orietation preset " << name << " is not stored.");
+  return false;
+}
+
+//----------------------------------------------------------------------------
+bool vtkMRMLSliceNode::RenameSliceOrientationPreset(const std::string &name, const std::string &updatedName)
+{
+  if (!name.compare("Reformat"))
+    {
+    vtkErrorMacro("RenameSliceOrientationPreset: Reformat has no orietation preset.");
+    return false;
+    }
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it = this->OrientationMatrices.find(name);
+  if (it != this->OrientationMatrices.end())
+    {
+    std::swap(this->OrientationMatrices[updatedName], it->second);
+    this->OrientationMatrices.erase(it);
```

Comment:
yes it is working, because the values of the map are all pointers. However, we decide to change the map to vector, right?